core/ibft.go

```diff
@@ -945,16 +962,31 @@ func (i *IBFT) buildProposal(ctx context.Context, view *proto.View) *proto.Messa
 	}
 
 	//	check the messages for any previous proposal (if they have any, it's the same proposal)
-	var previousProposal []byte
+	var (
+		previousProposal []byte
+		maxRound         uint64
+	)
 
+	// take previous proposal among the round change messages for the highest round
 	for _, msg := range rcc.RoundChangeMessages {
-		//	if message contains block, break
 		latestPC := messages.ExtractLatestPC(msg)
+		if latestPC == nil {
+			continue
+		}
+
+		// skip if message's round is equals to/less than maxRound
+		msgRound := msg.View.Round
+		if msgRound <= maxRound {
+			continue
+		}
 
-		if latestPC != nil {
-			previousProposal = messages.ExtractLastPreparedProposedBlock(msg)
+		lastPB := messages.ExtractLastPreparedProposedBlock(msg)
+		if lastPB == nil {
+			continue
+		}
 
-			break
+		if msgRound > maxRound {
```

Comment:
Is `maxRound = msgRound` needed here?

Reply:
Fixed aa164445f98b5c27a12ed590fc9102a690a399b8